compiler/rustc_const_eval/src/interpret/intern.rs

```diff
@@ -168,8 +168,48 @@ impl<'rt, 'mir, 'tcx: 'mir, M: CompileTimeMachine<'mir, 'tcx, const_eval::Memory
         mplace: &MPlaceTy<'tcx>,
         fields: impl Iterator<Item = InterpResult<'tcx, Self::V>>,
     ) -> InterpResult<'tcx> {
-        // ZSTs cannot contain pointers, so we can skip them.
-        if mplace.layout.is_zst() {
+        // We want to walk the aggregate to look for reference types to intern. While doing that we
+        // also need to take special care of interior mutability.
+        //
+        // As an optimization, however, if the allocation does not contain any pointers: we don't
+        // need to do the walk. It can be costly for big arrays for example (e.g. issue #93215).
+        let is_walk_needed = |mplace: &MPlaceTy<'tcx>| -> InterpResult<'tcx, bool> {
+            // ZSTs cannot contain pointers, we can avoid the interning walk.
+            if mplace.layout.is_zst() {
+                return Ok(false);
+            }
+
+            // Now, check whether this allocation contains reference types (as relocations).
+            //
+            // Note, this check may sometimes not be cheap, so we only do it when the walk we'd like
+            // to avoid could be expensive: on the potentially larger types, arrays and slices,
+            // rather than on all aggregates unconditionally.
+            if matches!(mplace.layout.ty.kind(), ty::Array(..) | ty::Slice(..)) {
+                let Some((size, align)) = self.ecx.size_and_align_of_mplace(&mplace)? else {
+                    // We do the walk if we can't determine the size of the mplace: we may be
+                    // dealing with extern types here in the future.
+                    return Ok(true);
+                };
+
+                // If there are no refs or relocations in this allocation, we can avoid the
+                // interning walk.
+                if let Some(alloc) = self.ecx.get_ptr_alloc(mplace.ptr, size, align)? {
+                    if !alloc.has_relocations() {
+                        return Ok(false);
+                    }
+                }
```

Comment:
A couple things here:

- I'm not using a let chain because rustfmt ignores it, and doesn't format this block if I do.
- if we fail to get an AllocRef, we will do the walk. I believe we could equally not do it, since there shouldn't be any relocations anyways, but it shouldn't matter in practice IIUC ?

Reply:
Yeah if `get_ptr_alloc` returns `None` then this is a ZST. So better to not recurse, that seems less confusing.

Reply:
I'll do that, then, thank you. 

Interestingly I think I saw some `None`s here that didn't apparently hit the ZST early return at the beginning of the function, would you find this unusual or completely fine ?

Reply:
That sounds odd. Basically it should be the case that if `get_ptr_alloc` returns `None` then `size` must be 0.

However I guess `size` being 0 does not mean it is a ZST... it could be an empty slice. Dynamically sized types are not considered ZST. So yeah I guess this is fine.